Add Guidance wrt Labelling to Naming and Rules Best Practices (#2691)
prometheus/docs: docs/practices/labels.md

---
title: Labels
sort_rank: 2
---

The label conventions presented in this document are not required
for using Prometheus, but can serve as both a style-guide and a collection of
best practices. Individual organizations may want to approach some of these
practices, e.g. naming conventions, differently.

## Labels

Prometheus labels can come from both the target and from

> [relabeling in discovery](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config) as well as from the target itself.

Reviewer: ```suggestion
Prometheus labels can come from both the target itself and from
[relabeling in discovery](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config).
```

By default Prometheus configures two primary discovery target labels.

- `job`

>   - The `job`  is a default target label set by the scrape configs and is used to identify metrics scraped from the same target/exporter.

Reviewer: ```suggestion
  - The `job` is a default target label set by the scrape configs and is used to identify metrics scraped by the same scrape config.
```

>   - If not specified in PromQL expressions, they will match unrelated metrics with the same name. This is especially true in a multi system or multi tenant installation

Reviewer: It might be the intention of the user to not distinguish between unrelated metrics, for example when aggregating across jobs. So I'd turn this into some positive statement instead (use "specify" instead of "not specified"). Multi system seems vague and I don't think multi tenancy has anything to do with this.

```suggestion
  - If specified in PromQL expressions, they will match metrics scraped by the same job.
```

Author: > It might be the intention of the user to not distinguish between unrelated metrics, for example when aggregating across jobs. So I'd turn this into some positive statement instead (use "specify" instead of "not specified").

I agree that the user may want to aggregate across all, or a subset of jobs. But to do so, they should be explicit with the `job` label values. But I expect unintended label stripping to vastly outnumber cross-job aggregation use cases.

> Multi system seems vague and I don't think multi tenancy has anything to do with this.

"Multi-tenent systems" may not be the best term, but I'm referring to a Prometheus, run as a platform for multiple teams (e.g by a DevEx or Platform Engineering team), to prevent every team running their own siloed Prometheus stack. In such a setup, all PromQL expressions should be scoped with a `job` label, to ensure the metrics are from the the expected exporters.

Or framed another way, in such a centralised stack, always write `up{job=bla}`, never `up{}`

> WARNING: When using `without`, be careful not to strip out the `job` label accidentally.

Reviewer: It might be intentional , so I think this needs to be conditional.

- `instance`

>   - The `instance` label will include the `ip:port` what was scraped

Reviewer: ```suggestion
  - The `instance` label will include the `ip:port` what was scraped, identifying the target instance.
```

> 

Reviewer: Don't you need a similar warning for "instance" , depending on usage?

Author: While `instance` is one of the standard scrape time labels, like `job`, stripping it doesn't have the same blast radius. Stripping `instance` will make metrics hard to debug, but should will still work.

For certain use cases that require using multiple layers of rules (e.g in a multi region, multi layered tree of Prometheus), you may want to strip out `instance` at the higher aggregation layers to manage label cardinality (e.g instance labels make sense to the per region aggregation, but can be problematic if aggregated at the global level)

I've added a warning that stripping `instance` can make it harder to debug scrape time issues with a metric though.
